bmc.c

```diff
@@ -126,6 +129,59 @@ int dscmp(struct dataset *a, struct dataset *b)
 	return diff < 0 ? A_BETTER : B_BETTER;
 }
 
+static enum port_state hsr_state_decision(struct clock *c, struct port *r,
+					  int (*compare)(struct dataset *a, struct dataset *b))
+{
+	struct dataset *clock_best, *port_best, *pair_best;
+	struct port *q;
+
+	q = port_get_paired(r);
+	clock_best = clock_best_foreign(c);
+	port_best = port_best_foreign(r);
+	pair_best = port_best_foreign(q);
```

Comment:
> Seems like these could end up being NULL.

This is handled in the `compare` function, which will be `dscmp`. If one is NULL it will simply prioritize the other port.